### PR TITLE
Optimize ldc.i8

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
Optimize Ldc.I8 calls when the value fits an int.

This saves 3 to 7 bytes of code.

e.g., this IL (9 bytes):
```
IL_0047:  ldc.i8 0x0
IL_0050:  ....
```
is optimized into this (2 bytes)
```
IL_0032:  ldc.i4.0 
IL_0033:  conv.i8 
IL_0034:  ....
```

This (9B):
```
IL_00b9:  ldc.i8 0x7f
IL_00c2:  .... 
```
into (3B):
```
IL_00a6:  ldc.i4.s 0x7f
IL_00a8:  conv.i8 
IL_00a9:  ....
```

And this (9B):
```
IL_0059:  ldc.i8 0x7fffffff
IL_0062:  ....
```
into (6B)
```
IL_003a:  ldc.i4 2147483647
IL_003f:  conv.i8 
IL_0040:  .... 
```

### CODE REVIEW

**Comment by jbevain** [comment]
Interesting!

It doesn't belong to OptimizeMacros, semantically.
We could have a Optimize() extension method which would optimize macros + do other optimizations, including this one. There's a few other optimizations that could applied (remove nops, remove br to next instr, etc).

**Comment by StephaneDelcroix** [comment]
it indeed doesn't belong to OptimizeMacro, but that was a way to include the capability without adding new API.

I'll gladly add `Optimize()`.

My list also contains replacing `Stloc X` followed directly by `LdLoc X` by `Dup` and `StLoc X`

**Comment by jbevain** [comment]
<3

**Comment by StephaneDelcroix** [comment]
there @jbevain, fixed

**Comment by jbevain** [comment]
Thanks! Please update the code to follow to mono coding conventions.

File: rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs [hunk 1]
```diff
@@ -174,6 +174,30 @@ static void MakeMacro (Instruction instruction, OpCode opcode)
 			instruction.Operand = null;
 		}
 
+		public static void Optimize(this MethodBody self)
```
Comment (jbevain) [inline rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs]: `Optimize (`

File: rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs [hunk 2]
```diff
@@ -174,6 +174,30 @@ static void MakeMacro (Instruction instruction, OpCode opcode)
...
+		{
+			if (self == null)
+				throw new ArgumentNullException("self");
```
Comment (jbevain) [inline rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs]: `ArgumentNullException ("self")`

File: rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs [hunk 3]
```diff
@@ -174,6 +174,30 @@ static void MakeMacro (Instruction instruction, OpCode opcode)
...
+
+			OptimizeLongs(self);
+			OptimizeMacros(self);
+		}
+
+		static void OptimizeLongs(this MethodBody self)
+		{
+			var method = self.Method;
+			for (var i = 0; i < self.Instructions.Count; i++) {
+				var instruction = self.Instructions[i];
```
Comment (jbevain) [inline rocks/Mono.Cecil.Rocks/MethodBodyRocks.cs]: `self.Instructions [i];`

**Comment by jbevain** [comment]
Thanks!

**Pull request merged**